### PR TITLE
added changed files that touch the core (minus derivation + contextfn)

### PR DESCRIPTION
**Pull request by reginaldlong**

### CODE REVIEW

File: src/edu/stanford/nlp/sempre/Parser.java [hunk 1]
```diff
@@ -56,6 +56,9 @@
 
     @Option(gloss = "Dump all features (for debugging)")
     public boolean dumpAllFeatures = false;
+
+    @Option(gloss = "Whether rules are marked as floating by default")
+    public boolean defaultIsFloating = true;
```
Comment (percyliang) [inline src/edu/stanford/nlp/sempre/Parser.java]: Why is this moved?

File: src/edu/stanford/nlp/sempre/Rule.java [hunk 2]
```diff
@@ -127,6 +127,23 @@ public boolean isAnchored() {
     else if (a == 0.0)
       return false;
     else
-      return f == 1.0 ? false : !FloatingParser.opts.defaultIsFloating;
+      return f == 1.0 ? false : !Parser.opts.defaultIsFloating;
   }
+
+  public boolean isSegmentRoot() {
```
Comment (percyliang) [inline src/edu/stanford/nlp/sempre/Rule.java]: Better to not add these here or else Rule.java is going to get bloated.  Add them as static methods whereever they are actually used.

File: src/edu/stanford/nlp/sempre/ParserState.java [hunk 3]
```diff
@@ -227,6 +227,58 @@ protected void pruneCell(String cellDescription, List<Derivation> derivations) {
     return derivs;
   }
 
+  public List<Derivation> gatherTokenAndPhraseDerivations(int spanStart, int spanEnd) {
```
Comment (percyliang) [inline src/edu/stanford/nlp/sempre/ParserState.java]: Seems like this only depends on Example - make this static function and pass in Example for clarity.  Document what this function does.

Comment (percyliang) [inline src/edu/stanford/nlp/sempre/ParserState.java]: Also, avoid duplicate code with previous function - refactor so things only show up once.  Probably the best is to have the above function call this one.

**Comment by percyliang** [comment]
@reggylong Left some basic comments.

File: src/edu/stanford/nlp/sempre/Parser.java (hunk 1 above)
Comment (reggylong) [inline src/edu/stanford/nlp/sempre/Parser.java]: It seems better to have it in one place, instead of in both the Floating and ShiftReduceParser

**Comment by reggylong** [comment]
@percyliang Take another look?

File: src/edu/stanford/nlp/sempre/Example.java [hunk 4]
```diff
@@ -37,6 +37,7 @@
   // What we should try to predict.
   @JsonProperty public Formula targetFormula;  // Logical form (e.g., database query)
   @JsonProperty public Value targetValue;  // Denotation (e.g., answer)
+  @JsonProperty public Value intermediateTargetValues; // Intermediate denotations
```
Comment (percyliang) [inline src/edu/stanford/nlp/sempre/Example.java]: Why is this a single Value rather than a list?

Comment (percyliang) [inline src/edu/stanford/nlp/sempre/Example.java]: If it's a list, make this a ListValue and document what are the elements.

File: src/edu/stanford/nlp/sempre/Parser.java (hunk 1 above)
Comment (percyliang) [inline src/edu/stanford/nlp/sempre/Parser.java]: Sure.  Document that this is only used by the individual parsers, and it doesn't affect the beam parser.

**Comment by percyliang** [comment]
LGTM modulo comments.  Also you need to resolve conflicts.

**Comment by reggylong** [comment]
Do I need to be a contributor to resolve the conflicts?

**Comment by percyliang** [comment]
No, I can merge once you've fixed what you need to.
